doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -0,0 +1,331 @@
+a minor release.
+
+A basic principle we will adhere to is "one function, one location". Functions
+that are exposed in more than one namespace (e.g., many functions are present
+in ``numpy`` and ``numpy.lib``) need to find a single home.
+
+We will reorganize the API reference guide along main and submodule namespaces,
+and only within the main namespace use the current subdivision along
+functionality groupings. Also by "mainstream" and special-purpose namespaces.
+Details TBD, something like:
+
+::
+
+    # Regular/recommended user-facing namespaces for general use. Present these
+    # as the primary set of namespaces to the users.
+    numpy
+    numpy.exceptions
+    numpy.fft
+    numpy.linalg
+    numpy.ma
+    numpy.polynomial
+    numpy.random
+    numpy.testing
+    numpy.typing
+
+    # Special-purpose namespaces. Keep these, but document them in a separate
+    # grouping in the reference guide and explain their purpose.
+    numpy.array_api
+    numpy.ctypeslib
+    numpy.emath
+    numpy.f2py  # only a couple of public functions, like `compile` and `get_include`
+    numpy.math
+    numpy.lib.stride_tricks
+    numpy.rec
+    numpy.types
+
+    # Legacy (prefer not to use). Third grouping in the reference guide.
+    numpy.char
+    numpy.distutils
+    numpy.matrixlib  # or deprecate?
+
+    # To remove
+    numpy.compat
+    numpy.core  # rename to _core
+    numpy.doc
+    numpy.matlib
+    numpy.version
+
+    # To clean out or somehow deal with: everything in `numpy.lib`
+
+.. note::
+
+    TBD: will we preserve ``np.lib`` or not? It only has a couple of unique
+    functions/objects, like ``Arrayterator`` (a candidate for removal) and the
+    ``stride_tricks`` subsubmodule. ``numpy.lib`` itself is not a coherent
+    namespace, and does not even have a reference guide page.
+
+.. note::
+
+   TBD: is ``numpy.math`` actually used at all? If not, should we hide/remove
+   it?
+
+We will make all submodules available lazily, so that users don't have to type
+``import numpy.xxx`` but can use ``import numpy as np; np.xxx.*``, while at the
+same time not negatively impacting the overhead of ``import numpy``. This has
+been very helpful for teaching scikit-image and SciPy.
+
+
+Reducing the number of ways to select dtypes
+````````````````````````````````````````````
+
+The many dtype classes, instances, aliases and ways to select them are one of
+the larger usability problems in the NumPy API. E.g.:
+
+.. code:: python
+
+   >>> # np.intp is different, but compares equal too
+   >>> np.int64 == np.int_ == np.dtype('i8') == np.sctypeDict['i8']
+   True
+   >>> np.float64 == np.double == np.float_ == np.dtype('f8') == np.sctypeDict['f8']
+   True
+   ### Really?
+   >>> np.clongdouble == np.clongfloat == np.longcomplex == np.complex256
+   True
+
+These aliases can go: https://numpy.org/devdocs/reference/arrays.scalars.html#other-aliases
+
+To discuss:
+
+- move *all* dtype-related classes to ``np.types``?
```

Comment:
I think that makes sense, though would love to stop using `type` as equivalent to a data type as much as possible... Since this is a new module, there is an opportunity in naming it.  `data_types` or `datatypes` are probably too long; is `dtypes` an option? Or too confusing with `np.dtype`?

Or could it be `np.dt`?  One could then think of pre-defining some `dtype` so that one could do `np.array(..., dtype=np.dt.f8)`.

Reply:
Thanks for the thoughts! That is gh-23358 rather than this NEP, here we should just follow what's decided there.

Reply:
OK, resolving here then.